lib/IRGen/GenClass.h

```diff
@@ -75,6 +75,45 @@ namespace irgen {
     HasUpdateCallback = true
   };
 
+  /// Creates a layout for the class \p classType with allocated tail elements
+  /// \p tailTypes.
+  ///
+  /// The caller is responsible for deleting the returned StructLayout.
+  StructLayout *getClassLayoutWithTailElems(IRGenModule &IGM, SILType classType,
+                                            llvm::ArrayRef<SILType> tailTypes);
+
+  ClassDecl *getRootClassForMetaclass(IRGenModule &IGM, ClassDecl *theClass);
+
+  enum class ClassMetadataStrategy {
+    /// Does the given class have resilient ancestry, or is the class itself
+    /// generic?
+    Resilient,
+
+    /// Does the class require at in-place initialization because of
+    /// non-fixed size properties or generic ancestry? The class does not
+    /// export a static symbol visible to Objective-C code.
+    Singleton,
```

Comment:
I'm confused why you named this one "Singleton". I guess you mean "non-generic", but I think "NonGeneric" or "RuntimeNonGeneric" would be better.

Alternately: Resilient, SwiftNonGeneric, ObjCUpdate, ObjCFixedOrUpdate, ObjCFixed.

Reply:
The name comes from the runtime’s “SingletonMetadataInitialization” and related code, which is used for non-generic resilient types. ObjCFixed is not really accurate, because it’s used for non-objc classes without any dynamic initialization as well.